fine-grained control systemd to start/stop/restart ceph services at once

david-z opened the pull request:
http://tracker.ceph.com/issues/13497

This will be more convenient to start/stop/restart each kind of ceph services itself at once even if multiple services are running on the same machine.

Fixes: #13497
Signed-off-by: Zhi Zhang zhangz.david@outlook.com

liewegas commented:
Would it make more sense for ceph-osd.target to be PartOf ceph.target, and ceph-osd.service to be PartOf ceph-osd.target?  That way when you start ceph.target you'll see ceph-osd.target is started, and so on...

david-z commented:
@liewegas  Thanks, Sage. This makes sense and hierarchical. Please review the new changes.

liewegas commented:
LGTM, thanks!

@ktdreyer care to take a look?

ktdreyer commented:
LGTM; @smithfarm @branto1 care to review?

branto1 commented:
Mostly lgtm, my only concern is that we might make disabling ceph too complicated if we make all the ceph-*.target files WantedBy multi-user.target.

liewegas commented:
Those are redundant if ceph.target is WantedBy multi-user.target, right?  In that case let's drop them...

smithfarm commented:
I have reviewed this PR. It does seem to me that all the lines that say 

```
WantedBy=multi-user.target ceph.target
```

should be changed to:

```
WantedBy=ceph.target
```

But even then I would be reluctant to merge this without any testing.

david-z commented:
@liewegas @smithfarm @branto1 

If we drop "WantedBy multi-user.target" parts from each ceph-*.target file, then we can't start/stop/restart each kind of ceph services itself at once even if multiple ceph services are running on the same machine.

For example, we drop "WantedBy multi-user.target" parts and there are 3 OSDs whose ids are not successive, such as osd.1, osd.4, osd.6, and 1 MDS, such as mds.1, running on one machine.

If we want to restart all the OSDs at once, we need to run something like this
'''
sudo systemctl restart ceph-osd@1
sudo systemctl restart ceph-osd@4
sudo systemctl restart ceph-osd@6
''' 

It is not working if we run following cmd to restart them at once
'''
sudo systemctl restart ceph-osd.target
'''

But if we have "WantedBy multi-user.target" in each ceph-*.target file, we can do above both ways according to different situation.

zphj1987 commented:
[lab8106][INFO  ] Running command: systemctl enable ceph.target
[lab8106][INFO  ] Running command: systemctl enable ceph-mon@lab8106
[lab8106][WARNIN] Failed to issue method call: No such file or directory
[lab8106][ERROR ] RuntimeError: command returned non-zero exit status: 1
[ceph_deploy.mon][ERROR ] Failed to execute command: systemctl enable ceph-mon@lab8106
[ceph_deploy][ERROR ] GenericError: Failed to create 1 monitors

[root@lab8106 ceph]# systemctl enable ceph-mon@lab8106
Failed to issue method call: No such file or directory


centos 7 ,ceph version 10.0.2 (86764eaebe1eda943c59d7d784b893ec8b0c6ff9)
deploy faild

david-z commented:
@zphj1987 

I just tried ceph version 10.0.1 with ceph-deploy version 1.5.28. I can deploy ceph successfully. Could you pls give me following information?

1. which ceph-deploy version are you using?

2. could you pls check if your monitor host contains following files after installing ceph package?
[ceph@xxx ~]$ ls -l /lib/systemd/system/ceph*    
-rw-r--r-- 1 root root 317 Jan 13 12:17 /lib/systemd/system/ceph-create-keys@.service
-rw-r--r-- 1 root root 205 Jan 13 12:17 /lib/systemd/system/ceph-disk@.service
-rw-r--r-- 1 root root 162 Jan 13 12:17 /lib/systemd/system/ceph-mds.target
-rw-r--r-- 1 root root 432 Jan 13 12:17 /lib/systemd/system/ceph-mds@.service
-rw-r--r-- 1 root root 162 Jan 13 12:17 /lib/systemd/system/ceph-mon.target
-rw-r--r-- 1 root root 659 Jan 13 12:17 /lib/systemd/system/ceph-mon@.service
-rw-r--r-- 1 root root 162 Jan 13 12:17 /lib/systemd/system/ceph-osd.target
-rw-r--r-- 1 root root 543 Jan 13 12:17 /lib/systemd/system/ceph-osd@.service
-rw-r--r-- 1 root root 166 Jan 13 12:17 /lib/systemd/system/ceph-radosgw.target
-rw-r--r-- 1 root root 404 Jan 13 12:17 /lib/systemd/system/ceph-radosgw@.service
-rw-r--r-- 1 root root 128 Jan 13 12:17 /lib/systemd/system/ceph.target

3. run following command on your monitor host to check if your monitor is there.  
[ceph@xxx ~]$ sudo systemctl list-dependencies ceph-mon.target
ceph-mon.target
|-ceph-mon@xxx.service

david-z commented:
@zphj1987 

I just tried latest master (10.0.2) with ceph-deploy 1.5.28. I can also deploy ceph successfully.

zphj1987 commented:
1.ceph version
[root@lab8106 ceph]# rpm -qa|grep ceph
python-cephfs-10.0.2-0.el7.x86_64
ceph-10.0.2-0.el7.x86_64
ceph-deploy-1.5.31-0.noarch
ceph-test-10.0.2-0.el7.x86_64
ceph-common-10.0.2-0.el7.x86_64
libcephfs1-10.0.2-0.el7.x86_64
ceph-selinux-10.0.2-0.el7.x86_64
2.the file
[root@lab8106 ceph]# ls -l /lib/systemd/system/ceph*
-rw-r--r-- 1 root root 317 Jan 14 01:33
/lib/systemd/system/ceph-create-keys@.service
-rw-r--r-- 1 root root 205 Jan 14 01:33 /lib/systemd/system/ceph-disk@
.service
-rw-r--r-- 1 root root 432 Jan 14 01:33 /lib/systemd/system/ceph-mds@
.service
-rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-mds.target
-rw-r--r-- 1 root root 659 Jan 14 01:33 /lib/systemd/system/ceph-mon@
.service
-rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-mon.target
-rw-r--r-- 1 root root 543 Jan 14 01:33 /lib/systemd/system/ceph-osd@
.service
-rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-osd.target
-rw-r--r-- 1 root root 404 Jan 14 01:33 /lib/systemd/system/ceph-radosgw@
.service
-rw-r--r-- 1 root root 166 Jan 14 01:33
/lib/systemd/system/ceph-radosgw.target
-rw-r--r-- 1 root root 128 Jan 14 01:33 /lib/systemd/system/ceph.target
3.the command
sudo systemctl list-dependencies ceph-mon.target

4、ceph-deploy version
ceph-deploy-1.5.31-0.noarch

so you see
[root@lab8106 ceph]# systemctl start ceph-mon@lab8106
[root@lab8106 ceph]# systemctl enable ceph-mon@lab8106
Failed to issue method call: No such file or directory

systemctl start ceph-mon@lab8106    command work well
 systemctl enable ceph-mon@lab8106 went wrong

ceph-deploy when create a mon or osd  ,it will do systemctl enable
ceph-mon@lab8106   so  it went wrong







2016-01-18 14:07 GMT+08:00 Zhi Zhang <notifications@github.com>:

> @zphj1987 <https://github.com/zphj1987>
>
> I just tried latest master (10.0.2) with ceph-deploy 1.5.28. I can also
> deploy ceph successfully.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/6276#issuecomment-172433935>.
>

zphj1987 commented:
[lab8106][WARNIN] DEBUG:ceph-disk:Moving mount to final location...
[lab8106][WARNIN] INFO:ceph-disk:Running command: /bin/mount -o
noatime,inode64 -- /dev/sdd1 /var/lib/ceph/osd/ceph-2
[lab8106][WARNIN] INFO:ceph-disk:Running command: /bin/umount -l --
/var/lib/ceph/tmp/mnt.q566wa
[lab8106][WARNIN] DEBUG:ceph-disk:Starting ceph osd.2...
[lab8106][WARNIN] INFO:ceph-disk:Running command: /usr/bin/systemctl enable
ceph-osd@2
[lab8106][WARNIN] Failed to issue method call: No such file or directory
[lab8106][WARNIN] Traceback (most recent call last):
[lab8106][WARNIN]   File "/usr/sbin/ceph-disk", line 4036, in <module>
[lab8106][WARNIN]     main(sys.argv[1:])
[lab8106][WARNIN]   File "/usr/sbin/ceph-disk", line 3990, in main
[lab8106][WARNIN]     args.func(args)
[lab8106][WARNIN]   File "/usr/sbin/ceph-disk", line 2561, in main_activate
[lab8106][WARNIN]     osd_id=osd_id,
[lab8106][WARNIN]   File "/usr/sbin/ceph-disk", line 2134, in start_daemon
[lab8106][WARNIN]     raise Error('ceph osd start failed', e)
[lab8106][WARNIN] __main__.Error: Error: ceph osd start failed: Command
'['/usr/bin/systemctl', 'enable', 'ceph-osd@2']' returned non-zero exit
status 1
[lab8106][ERROR ] RuntimeError: command returned non-zero exit status: 1
[ceph_deploy][ERROR ] RuntimeError: Failed to execute command: ceph-disk -v
activate --mark-init systemd --mount /dev/sdd1



i do it again  ceph-deploy-1.5.28

you can see  /usr/bin/systemctl enable ceph-osd@2
went wrong

2016-01-20 18:04 GMT+08:00 张鹏 <zphj1987@gmail.com>:

> 1.ceph version
> [root@lab8106 ceph]# rpm -qa|grep ceph
> python-cephfs-10.0.2-0.el7.x86_64
> ceph-10.0.2-0.el7.x86_64
> ceph-deploy-1.5.31-0.noarch
> ceph-test-10.0.2-0.el7.x86_64
> ceph-common-10.0.2-0.el7.x86_64
> libcephfs1-10.0.2-0.el7.x86_64
> ceph-selinux-10.0.2-0.el7.x86_64
> 2.the file
> [root@lab8106 ceph]# ls -l /lib/systemd/system/ceph*
> -rw-r--r-- 1 root root 317 Jan 14 01:33
> /lib/systemd/system/ceph-create-keys@.service
> -rw-r--r-- 1 root root 205 Jan 14 01:33 /lib/systemd/system/ceph-disk@
> .service
> -rw-r--r-- 1 root root 432 Jan 14 01:33 /lib/systemd/system/ceph-mds@
> .service
> -rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-mds.target
> -rw-r--r-- 1 root root 659 Jan 14 01:33 /lib/systemd/system/ceph-mon@
> .service
> -rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-mon.target
> -rw-r--r-- 1 root root 543 Jan 14 01:33 /lib/systemd/system/ceph-osd@
> .service
> -rw-r--r-- 1 root root 162 Jan 14 01:33 /lib/systemd/system/ceph-osd.target
> -rw-r--r-- 1 root root 404 Jan 14 01:33 /lib/systemd/system/ceph-radosgw@
> .service
> -rw-r--r-- 1 root root 166 Jan 14 01:33
> /lib/systemd/system/ceph-radosgw.target
> -rw-r--r-- 1 root root 128 Jan 14 01:33 /lib/systemd/system/ceph.target
> 3.the command
> sudo systemctl list-dependencies ceph-mon.target
>
> 4、ceph-deploy version
> ceph-deploy-1.5.31-0.noarch
>
> so you see
> [root@lab8106 ceph]# systemctl start ceph-mon@lab8106
> [root@lab8106 ceph]# systemctl enable ceph-mon@lab8106
> Failed to issue method call: No such file or directory
>
> systemctl start ceph-mon@lab8106    command work well
>  systemctl enable ceph-mon@lab8106 went wrong
>
> ceph-deploy when create a mon or osd  ,it will do systemctl enable
> ceph-mon@lab8106   so  it went wrong
>
>
>
>
>
>
>
> 2016-01-18 14:07 GMT+08:00 Zhi Zhang <notifications@github.com>:
>
>> @zphj1987 <https://github.com/zphj1987>
>>
>> I just tried latest master (10.0.2) with ceph-deploy 1.5.28. I can also
>> deploy ceph successfully.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ceph/ceph/pull/6276#issuecomment-172433935>.
>>
>
>

zphj1987 commented:
i run it  on  centos 7

david-z commented:
@zphj1987 

It is really odd. I also run it on centos-7 without any problem.

[ceph@xxx ~]$ uname -a
Linux xxx 3.10.90-1-xxxx-xxxx.xx #1 SMP Fri Nov 13 10:00:42 CST 2015 x86_64 x86_64 x86
_64 GNU/Linux
[ceph@cxxx ~]$ systemctl --version
systemd 208
+PAM +LIBWRAP +AUDIT +SELINUX +IMA +SYSVINIT +LIBCRYPTSETUP +GCRYPT +ACL +XZ

For better debugging this issue, can you just try ceph-disk to create an OSD? For example,

[ceph@xxx ~]$ sudo ceph-disk -v zap /dev/sdg
[ceph@xxx ~]$ sudo ceph-disk -v prepare /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-allows-journal -i 0 --cluster cep
h
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-wants-journal -i 0 --cluster ceph
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-needs-journal -i 0 --cluster ceph
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph --show-config-value=fsid
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
mkfs_type
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
fs_type
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
mkfs_options_xfs
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
fs_mkfs_options_xfs
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
mount_options_xfs
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph --show-config-value=osd_jo
urnal_size
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
cryptsetup_parameters
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
dmcrypt_key_size
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
dmcrypt_type
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
INFO:ceph-disk:Will colocate journal with data on /dev/sdg
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:Creating journal partition num 2 size 10000 on /dev/sdg
INFO:ceph-disk:Running command: /sbin/sgdisk --new=2:0:10000M --change-name=2:ceph journal
--partition-guid=2:c2ee3319-9982-4e51-a506-cf7023b581fc --typecode=2:45b0969e-9b03-4f30-b4c
6-b4b80ceff106 --mbrtogpt -- /dev/sdg
Information: Moved requested sector from 34 to 2048 in
order to align on 2048-sector boundaries.
The operation has completed successfully.
DEBUG:ceph-disk:Calling partprobe on prepared device /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
DEBUG:ceph-disk:Journal is GPT partition /dev/disk/by-partuuid/c2ee3319-9982-4e51-a506-cf70
23b581fc
DEBUG:ceph-disk:Journal is GPT partition /dev/disk/by-partuuid/c2ee3319-9982-4e51-a506-cf70
23b581fc
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:Creating osd partition on /dev/sdg
INFO:ceph-disk:Running command: /sbin/sgdisk --largest-new=1 --change-name=1:ceph data --pa
rtition-guid=1:8b20d79e-590b-41cf-8449-f5b78604e491 --typecode=1:89c57f98-2fe5-4dc0-89c1-f3
ad0ceff2be -- /dev/sdg
Information: Moved requested sector from 20480001 to 20482048 in
order to align on 2048-sector boundaries.
The operation has completed successfully.
DEBUG:ceph-disk:Calling partprobe on created device /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
DEBUG:ceph-disk:Creating xfs fs on /dev/sdg1
INFO:ceph-disk:Running command: /sbin/mkfs -t xfs -f -i size=2048 -- /dev/sdg1
...
DEBUG:ceph-disk:Mounting /dev/sdg1 on /var/lib/ceph/tmp/mnt.DxL33t with options rw,noatime,
inode64
INFO:ceph-disk:Running command: /usr/bin/mount -t xfs -o rw,noatime,inode64 -- /dev/sdg1 /v
ar/lib/ceph/tmp/mnt.DxL33t
INFO:ceph-disk:Running command: /sbin/restorecon /var/lib/ceph/tmp/mnt.DxL33t
DEBUG:ceph-disk:Preparing osd data dir /var/lib/ceph/tmp/mnt.DxL33t
DEBUG:ceph-disk:Creating symlink /var/lib/ceph/tmp/mnt.DxL33t/journal -> /dev/disk/by-partu
uid/c2ee3319-9982-4e51-a506-cf7023b581fc
INFO:ceph-disk:Running command: /sbin/restorecon -R /var/lib/ceph/tmp/mnt.DxL33t/ceph_fsid.
33719.tmp
INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/tmp/mnt.DxL33t/ce
ph_fsid.33719.tmp
INFO:ceph-disk:Running command: /sbin/restorecon -R /var/lib/ceph/tmp/mnt.DxL33t/fsid.33719
.tmp
INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/tmp/mnt.DxL33t/fs
id.33719.tmp
INFO:ceph-disk:Running command: /sbin/restorecon -R /var/lib/ceph/tmp/mnt.DxL33t/journal_uu
id.33719.tmp
INFO:ceph-disk:Running command: /sbin/restorecon -R /var/lib/ceph/tmp/mnt.DxL33t/magic.3371
9.tmp
INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/tmp/mnt.DxL33t/ma
gic.33719.tmp
INFO:ceph-disk:Running command: /sbin/restorecon -R /var/lib/ceph/tmp/mnt.DxL33t
INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/tmp/mnt.DxL33t
DEBUG:ceph-disk:Unmounting /var/lib/ceph/tmp/mnt.DxL33t
INFO:ceph-disk:Running command: /bin/umount -- /var/lib/ceph/tmp/mnt.DxL33t
DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is /sys/dev/block/8:96/dm/uuid
INFO:ceph-disk:Running command: /sbin/sgdisk --typecode=1:4fbd7e29-9d25-41b8-afd0-062c0ceff
05d -- /dev/sdg
The operation has completed successfully.
DEBUG:ceph-disk:Calling partprobe on prepared device /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
INFO:ceph-disk:Running command: /usr/bin/udevadm trigger --action=add --sysname-match sdg1


After above steps, OSD should be started successfully. If not, you can try to figure out what is wrong with it and run following command.
 
[ceph@xxx ~]$ sudo ceph-disk -v activate /dev/sdg1
DEBUG:ceph-disk:get_dm_uuid /dev/sdg1 uuid path is /sys/dev/block/8:97/dm/uuid
INFO:ceph-disk:Running command: /sbin/blkid -o udev -p /dev/sdg1
INFO:ceph-disk:Running command: /sbin/blkid -p -s TYPE -o value -- /dev/sdg1
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup osd_
mount_options_xfs
DEBUG:ceph-disk:Mounting /dev/sdg1 on /var/lib/ceph/tmp/mnt.ku5qyK with options rw,noatime,
inode64
INFO:ceph-disk:Running command: /usr/bin/mount -t xfs -o rw,noatime,inode64 -- /dev/sdg1 /v
ar/lib/ceph/tmp/mnt.ku5qyK
INFO:ceph-disk:Running command: /sbin/restorecon /var/lib/ceph/tmp/mnt.ku5qyK
DEBUG:ceph-disk:Cluster uuid is 91f500d3-453a-4096-8504-4be0cac09b63
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph --show-config-value=fsid
DEBUG:ceph-disk:Cluster name is ceph
DEBUG:ceph-disk:OSD uuid is 8b20d79e-590b-41cf-8449-f5b78604e491
DEBUG:ceph-disk:OSD id is 20
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph --name=osd. --lookup init
INFO:ceph-disk:Running command: /usr/bin/ceph-detect-init --default sysvinit
DEBUG:ceph-disk:Marking with init system systemd
DEBUG:ceph-disk:ceph osd.20 data dir is ready at /var/lib/ceph/tmp/mnt.ku5qyK
INFO:ceph-disk:ceph osd.20 already mounted in position; unmounting ours.
DEBUG:ceph-disk:Unmounting /var/lib/ceph/tmp/mnt.ku5qyK
INFO:ceph-disk:Running command: /bin/umount -- /var/lib/ceph/tmp/mnt.ku5qyK
DEBUG:ceph-disk:Starting ceph osd.20...
INFO:ceph-disk:Running command: /usr/bin/systemctl enable ceph-osd@20
INFO:ceph-disk:Running command: /usr/bin/systemctl start ceph-osd@20


As you can see, I can both enable and start ceph-osd successfully. For your case, you might end up at enabling ceph-osd. Can you double check your centos kernel version and systemd version?

zphj1987 commented:
[root@lab8106 3.2_head]# uname -a
Linux lab8106 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015
x86_64 x86_64 x86_64 GNU/Linux
[root@lab8106 3.2_head]#  systemctl --version
systemd 208
+PAM +LIBWRAP +AUDIT +SELINUX +IMA +SYSVINIT +LIBCRYPTSETUP +GCRYPT +ACL +XZ

root@lab8106 3.2_head]# ceph-disk -v activate /dev/sdd1
DEBUG:ceph-disk:get_dm_uuid /dev/sdd1 uuid path is
/sys/dev/block/8:49/dm/uuid
DEBUG:ceph-disk:get_dm_uuid /dev/sdd1 uuid path is
/sys/dev/block/8:49/dm/uuid
INFO:ceph-disk:Running command: /usr/sbin/sgdisk -i 1 /dev/sdd
INFO:ceph-disk:Running command: /sbin/blkid -p -s TYPE -o value -- /dev/sdd1
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
--name=osd. --lookup osd_mount_options_xfs
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
--name=osd. --lookup osd_fs_mount_options_xfs
DEBUG:ceph-disk:Mounting /dev/sdd1 on /var/lib/ceph/tmp/mnt.0VadUF with
options noatime,inode64
INFO:ceph-disk:Running command: /usr/bin/mount -t xfs -o noatime,inode64 --
/dev/sdd1 /var/lib/ceph/tmp/mnt.0VadUF
INFO:ceph-disk:Running command: /usr/sbin/restorecon
/var/lib/ceph/tmp/mnt.0VadUF
DEBUG:ceph-disk:Cluster uuid is 8f8ad2ba-e9f5-4306-bb2f-2573f0959b3b
INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph
--show-config-value=fsid
DEBUG:ceph-disk:Cluster name is ceph
DEBUG:ceph-disk:OSD uuid is 8ede3814-fddd-4961-bced-62359c8ed68f
DEBUG:ceph-disk:OSD id is 2
INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
--name=osd. --lookup init
INFO:ceph-disk:Running command: /usr/bin/ceph-detect-init --default sysvinit
DEBUG:ceph-disk:Marking with init system systemd
DEBUG:ceph-disk:ceph osd.2 data dir is ready at /var/lib/ceph/tmp/mnt.0VadUF
INFO:ceph-disk:ceph osd.2 already mounted in position; unmounting ours.
DEBUG:ceph-disk:Unmounting /var/lib/ceph/tmp/mnt.0VadUF
INFO:ceph-disk:Running command: /bin/umount -- /var/lib/ceph/tmp/mnt.0VadUF
DEBUG:ceph-disk:Starting ceph osd.2...
INFO:ceph-disk:Running command: /usr/bin/systemctl enable ceph-osd@2
Failed to issue method call: No such file or directory
Traceback (most recent call last):
  File "/usr/sbin/ceph-disk", line 4036, in <module>
    main(sys.argv[1:])
  File "/usr/sbin/ceph-disk", line 3990, in main
    args.func(args)
  File "/usr/sbin/ceph-disk", line 2561, in main_activate
    osd_id=osd_id,
  File "/usr/sbin/ceph-disk", line 2134, in start_daemon
    raise Error('ceph osd start failed', e)
__main__.Error: Error: ceph osd start failed: Command
'['/usr/bin/systemctl', 'enable', 'ceph-osd@2']' returned non-zero exit
status 1


so   you see.  i don't know what different with you

2016-01-21 11:18 GMT+08:00 Zhi Zhang <notifications@github.com>:

> @zphj1987 <https://github.com/zphj1987>
>
> It is really odd. I also run it on centos-7 without any problem.
>
> [ceph@xxx ~]$ uname -a
> Linux xxx 3.10.90-1-xxxx-xxxx.xx #1 <https://github.com/ceph/ceph/pull/1>
> SMP Fri Nov 13 10:00:42 CST 2015 x86_64 x86_64 x86
> _64 GNU/Linux
> [ceph@cxxx ~]$ systemctl --version
> systemd 208
> +PAM +LIBWRAP +AUDIT +SELINUX +IMA +SYSVINIT +LIBCRYPTSETUP +GCRYPT +ACL
> +XZ
>
> For better debugging this issue, can you just try ceph-disk to create an
> OSD? For example,
>
> [ceph@xxx ~]$ sudo ceph-disk -v zap /dev/sdg
> [ceph@xxx ~]$ sudo ceph-disk -v prepare /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-allows-journal
> -i 0 --cluster cep
> h
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-wants-journal -i
> 0 --cluster ceph
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --check-needs-journal -i
> 0 --cluster ceph
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph
> --show-config-value=fsid
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> mkfs_type
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> fs_type
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> mkfs_options_xfs
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> fs_mkfs_options_xfs
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> mount_options_xfs
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph
> --show-config-value=osd_jo
> urnal_size
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> cryptsetup_parameters
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> dmcrypt_key_size
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> dmcrypt_type
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> INFO:ceph-disk:Will colocate journal with data on /dev/sdg
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:Creating journal partition num 2 size 10000 on /dev/sdg
> INFO:ceph-disk:Running command: /sbin/sgdisk --new=2:0:10000M
> --change-name=2:ceph journal
> --partition-guid=2:c2ee3319-9982-4e51-a506-cf7023b581fc
> --typecode=2:45b0969e-9b03-4f30-b4c
> 6-b4b80ceff106 --mbrtogpt -- /dev/sdg
> Information: Moved requested sector from 34 to 2048 in
> order to align on 2048-sector boundaries.
> The operation has completed successfully.
> DEBUG:ceph-disk:Calling partprobe on prepared device /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> DEBUG:ceph-disk:Journal is GPT partition
> /dev/disk/by-partuuid/c2ee3319-9982-4e51-a506-cf70
> 23b581fc
> DEBUG:ceph-disk:Journal is GPT partition
> /dev/disk/by-partuuid/c2ee3319-9982-4e51-a506-cf70
> 23b581fc
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:Creating osd partition on /dev/sdg
> INFO:ceph-disk:Running command: /sbin/sgdisk --largest-new=1
> --change-name=1:ceph data --pa
> rtition-guid=1:8b20d79e-590b-41cf-8449-f5b78604e491
> --typecode=1:89c57f98-2fe5-4dc0-89c1-f3
> ad0ceff2be -- /dev/sdg
> Information: Moved requested sector from 20480001 to 20482048 in
> order to align on 2048-sector boundaries.
> The operation has completed successfully.
> DEBUG:ceph-disk:Calling partprobe on created device /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> DEBUG:ceph-disk:Creating xfs fs on /dev/sdg1
> INFO:ceph-disk:Running command: /sbin/mkfs -t xfs -f -i size=2048 --
> /dev/sdg1
> ...
> DEBUG:ceph-disk:Mounting /dev/sdg1 on /var/lib/ceph/tmp/mnt.DxL33t with
> options rw,noatime,
> inode64
> INFO:ceph-disk:Running command: /usr/bin/mount -t xfs -o
> rw,noatime,inode64 -- /dev/sdg1 /v
> ar/lib/ceph/tmp/mnt.DxL33t
> INFO:ceph-disk:Running command: /sbin/restorecon
> /var/lib/ceph/tmp/mnt.DxL33t
> DEBUG:ceph-disk:Preparing osd data dir /var/lib/ceph/tmp/mnt.DxL33t
> DEBUG:ceph-disk:Creating symlink /var/lib/ceph/tmp/mnt.DxL33t/journal ->
> /dev/disk/by-partu
> uid/c2ee3319-9982-4e51-a506-cf7023b581fc
> INFO:ceph-disk:Running command: /sbin/restorecon -R
> /var/lib/ceph/tmp/mnt.DxL33t/ceph_fsid.
> 33719.tmp
> INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph
> /var/lib/ceph/tmp/mnt.DxL33t/ce
> ph_fsid.33719.tmp
> INFO:ceph-disk:Running command: /sbin/restorecon -R
> /var/lib/ceph/tmp/mnt.DxL33t/fsid.33719
> .tmp
> INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph
> /var/lib/ceph/tmp/mnt.DxL33t/fs
> id.33719.tmp
> INFO:ceph-disk:Running command: /sbin/restorecon -R
> /var/lib/ceph/tmp/mnt.DxL33t/journal_uu
> id.33719.tmp
> INFO:ceph-disk:Running command: /sbin/restorecon -R
> /var/lib/ceph/tmp/mnt.DxL33t/magic.3371
> 9.tmp
> INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph
> /var/lib/ceph/tmp/mnt.DxL33t/ma
> gic.33719.tmp
> INFO:ceph-disk:Running command: /sbin/restorecon -R
> /var/lib/ceph/tmp/mnt.DxL33t
> INFO:ceph-disk:Running command: /usr/bin/chown -R ceph:ceph
> /var/lib/ceph/tmp/mnt.DxL33t
> DEBUG:ceph-disk:Unmounting /var/lib/ceph/tmp/mnt.DxL33t
> INFO:ceph-disk:Running command: /bin/umount -- /var/lib/ceph/tmp/mnt.DxL33t
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg uuid path is
> /sys/dev/block/8:96/dm/uuid
> INFO:ceph-disk:Running command: /sbin/sgdisk
> --typecode=1:4fbd7e29-9d25-41b8-afd0-062c0ceff
> 05d -- /dev/sdg
> The operation has completed successfully.
> DEBUG:ceph-disk:Calling partprobe on prepared device /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> INFO:ceph-disk:Running command: /sbin/partprobe /dev/sdg
> INFO:ceph-disk:Running command: /usr/bin/udevadm settle --timeout=600
> INFO:ceph-disk:Running command: /usr/bin/udevadm trigger --action=add
> --sysname-match sdg1
>
> After above steps, OSD should be started successfully. If not, you can try
> to figure out what is wrong with it and run following command.
>
> [ceph@xxx ~]$ sudo ceph-disk -v activate /dev/sdg1
> DEBUG:ceph-disk:get_dm_uuid /dev/sdg1 uuid path is
> /sys/dev/block/8:97/dm/uuid
> INFO:ceph-disk:Running command: /sbin/blkid -o udev -p /dev/sdg1
> INFO:ceph-disk:Running command: /sbin/blkid -p -s TYPE -o value --
> /dev/sdg1
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup osd_
> mount_options_xfs
> DEBUG:ceph-disk:Mounting /dev/sdg1 on /var/lib/ceph/tmp/mnt.ku5qyK with
> options rw,noatime,
> inode64
> INFO:ceph-disk:Running command: /usr/bin/mount -t xfs -o
> rw,noatime,inode64 -- /dev/sdg1 /v
> ar/lib/ceph/tmp/mnt.ku5qyK
> INFO:ceph-disk:Running command: /sbin/restorecon
> /var/lib/ceph/tmp/mnt.ku5qyK
> DEBUG:ceph-disk:Cluster uuid is 91f500d3-453a-4096-8504-4be0cac09b63
> INFO:ceph-disk:Running command: /usr/bin/ceph-osd --cluster=ceph
> --show-config-value=fsid
> DEBUG:ceph-disk:Cluster name is ceph
> DEBUG:ceph-disk:OSD uuid is 8b20d79e-590b-41cf-8449-f5b78604e491
> DEBUG:ceph-disk:OSD id is 20
> INFO:ceph-disk:Running command: /usr/bin/ceph-conf --cluster=ceph
> --name=osd. --lookup init
> INFO:ceph-disk:Running command: /usr/bin/ceph-detect-init --default
> sysvinit
> DEBUG:ceph-disk:Marking with init system systemd
> DEBUG:ceph-disk:ceph osd.20 data dir is ready at
> /var/lib/ceph/tmp/mnt.ku5qyK
> INFO:ceph-disk:ceph osd.20 already mounted in position; unmounting ours.
> DEBUG:ceph-disk:Unmounting /var/lib/ceph/tmp/mnt.ku5qyK
> INFO:ceph-disk:Running command: /bin/umount -- /var/lib/ceph/tmp/mnt.ku5qyK
> DEBUG:ceph-disk:Starting ceph osd.20...
> INFO:ceph-disk:Running command: /usr/bin/systemctl enable ceph-osd@20
> INFO:ceph-disk:Running command: /usr/bin/systemctl start ceph-osd@20
>
> As you can see, I can both enable and start ceph-osd successfully. For
> your case, you might end up at enabling ceph-osd. Can you double check your
> centos kernel version and systemd version?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/6276#issuecomment-173438242>.
>

david-z commented:
Can you paste the output of preparing an OSD? Usually OSD will be started at preparing step.

yunlion commented:
@zphj1987，I've had the same problem，even deployed by ceph-deploy.so do you have solve it ?

zphj1987 commented:
@yunlion
after failed  ，just   change the enable to start  , the daemon will start
ok

2016-01-25 16:58 GMT+08:00 yunlion <notifications@github.com>:

> @zphj1987 <https://github.com/zphj1987>，I've had the same problem，even
> deployed by ceph-deploy.so do you have solve it ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/6276#issuecomment-174441160>.
>
